### PR TITLE
Improve punishment function

### PR DESCRIPTION
**Pull request by WeiLoy**

### CODE REVIEW

File: x/plugin/slashing_plugin.go [hunk 1]
```diff
@@ -50,10 +53,30 @@ const (
...
+}
+
+func (w *WaitSlashingNode) String() string {
+	v, err := json.Marshal(w)
+	if err != nil {
+		panic(err)
+	}
+	return string(v)
+}
+
```
Comment (GavinXu520) [inline x/plugin/slashing_plugin.go]: Recommended using that fmt.Sprintf() replace json.Marshal()

File: x/plugin/slashing_plugin.go [hunk 2]
```diff
@@ -113,52 +136,71 @@ func (sp *SlashingPlugin) BeginBlock(blockHash common.Hash, header *types.Header
 
 			slashQueue := make(staking.SlashQueue, 0)
```
Comment (GavinXu520) [inline x/plugin/slashing_plugin.go]: slashQueue := make(staking.SlashQueue, 0)  maybe move into ` if currentVersion == 0 || currentVersion >= uint32(0<<16|9<<11|0) { } else { // TODO Here}`

File: x/plugin/staking_plugin.go [hunk 3]
```diff
@@ -2552,12 +2552,19 @@ func calcDelegateIncome(epoch uint64, del *staking.Delegation, per []*reward.Del
 	if uint64(del.DelegateEpoch) == epoch {
 		return nil
 	}
+	// When the settlement period when the first delegation is the same as the settlement period when the node is staking,
+	// and when the delegation is performed again in the next settlement cycle,
+	// the "per" value of the node is not available at this time, so you need to directly lazy
```
Comment (bglmmz) [inline x/plugin/staking_plugin.go]: the comments should be more clearer.

Comment (WeiLoy) [reply x/plugin/staking_plugin.go]: It has been modified

File: x/plugin/slashing_plugin.go (hunk 2 above)
Comment (WeiLoy) [reply x/plugin/slashing_plugin.go]: It has been modified

**Pull request merged**